rust/otap-dataflow/crates/engine/src/exporter.rs

```diff
@@ -215,50 +467,310 @@ mod tests {
+                }
+
+                // Send a Config event.
+                ctx.send_config(Value::Null)
                     .await
-                    .expect("Failed to send TimerTick");
-                ctx.sleep(Duration::from_millis(50)).await;
-            }
+                    .expect("Failed to send Config");
+
+                // Send a data message
+                ctx.send_pdata(TestMsg("Hello Exporter".into()))
+                    .await
+                    .expect("Failed to send data message");
+
+                // Send shutdown
+                ctx.send_shutdown(Duration::from_millis(200), "test complete")
+                    .await
+                    .expect("Failed to send Shutdown");
+            })
+        }
+    }
+
+    /// Validation closure that checks the expected counter values
+    fn validation_procedure()
+    -> impl FnOnce(TestContext<TestMsg>) -> std::pin::Pin<Box<dyn Future<Output = ()>>> {
+        |ctx| {
+            Box::pin(async move {
+                ctx.counters().assert(
+                    3, // timer tick
+                    1, // message
+                    1, // config
+                    1, // shutdown
+                );
+            })
+        }
+    }
+
+    #[test]
+    fn test_exporter_with_not_send_effect_handler() {
+        let test_runtime = TestRuntime::new();
+        let exporter = ExporterWrapper::with_not_send(
+            ExporterWithNotSendEffectHandler::without_send_test(test_runtime.counters()),
+            test_runtime.config(),
+        );
+
+        test_runtime
+            .set_exporter(exporter)
+            .run_test(scenario())
+            .run_validation(validation_procedure());
+    }
+
+    #[test]
+    fn test_exporter_with_send_effect_handler() {
+        let test_runtime = TestRuntime::new();
+        let exporter = ExporterWrapper::with_send(
+            ExporterWithSendEffectHandler::with_send_test(
+                test_runtime.counters(),
+                |effect_handler| {
+                    exec_in_send_env(|| {
+                        _ = effect_handler.exporter_name();
+                    });
+                },
+            ),
+            test_runtime.config(),
+        );
+
+        test_runtime
+            .set_exporter(exporter)
+            .run_test(scenario())
+            .run_validation(validation_procedure());
+    }
+
+    fn make_chan() -> (
+        mpsc::Sender<ControlMsg>,
+        mpsc::Sender<String>,
+        MessageChannel<String>,
+    ) {
+        let (control_tx, control_rx) = mpsc::Channel::<ControlMsg>::new(10);
+        let (pdata_tx, pdata_rx) = mpsc::Channel::<String>::new(10);
+        (
+            control_tx,
+            pdata_tx,
+            MessageChannel::new(control_rx, pdata_rx),
+        )
+    }
+
+    #[tokio::test]
+    async fn test_control_priority() {
+        let (control_tx, pdata_tx, mut channel) = make_chan();
+
+        control_tx
```

Comment:
Since we want to test that control messages are prioritized over data messages, we should flip the order here and emit a data message first and then emit a control message.

Reply:
Done